src/rpc/rpc.cc

```diff
@@ -295,5 +299,129 @@ DGL_REGISTER_GLOBAL("distributed.server_state._CAPI_DGLRPCGetServerState")
+  int client_id = args[4];
+  int service_id = args[5];
+  int msg_seq = args[6];
+  std::string pickle_data = args[7];
+  NDArray ID = args[8];
+  NDArray part_id = args[9];
+  NDArray g2l_id = args[10];
+  NDArray local_data = args[11];
+  // Data
+  int64_t ID_size = ID.GetSize() / sizeof(int64_t);
+  int64_t* ID_data = static_cast<int64_t*>(ID->data);
+  int64_t* part_id_data = static_cast<int64_t*>(part_id->data);
+  int64_t* g2l_id_data = static_cast<int64_t*>(g2l_id->data);
+  char* local_data_char = static_cast<char*>(local_data->data);
+  std::vector<int64_t> local_ids;
+  std::vector<int64_t> local_ids_orginal;
+  std::vector<int64_t> local_data_shape;
+  std::vector<std::vector<int64_t> > remote_ids(machine_count);
+  std::vector<std::vector<int64_t> > remote_ids_original(machine_count);
+  // Get row size (in bytes)
+  int row_size = 1;
+  for (int i = 0; i < local_data->ndim; ++i) {
+    local_data_shape.push_back(local_data->shape[i]);
+    if (i != 0) {
+      row_size *= local_data->shape[i];
+    }
+  }
+  row_size *= (local_data->dtype.bits / 8);
+  CHECK_GT(local_data_shape.size(), 0);
+  CHECK_EQ(row_size * local_data_shape[0], local_data.GetSize());
+  // Get local id (used in local machine) and
+  // remote id (send to remote machine)
+  int64_t idx = 0;
+  for (int64_t i = 0; i < ID_size; ++i) {
+    int64_t p_id = part_id_data[i];
+    if (p_id == local_machine_id) {
+      int64_t l_id = g2l_id_data[idx++];
+      local_ids.push_back(l_id);
+      local_ids_orginal.push_back(i);
+    } else {
+      int64_t id = ID_data[i];
+      remote_ids[p_id].push_back(id);
+      remote_ids_original[p_id].push_back(i);
+    }
+  }
+  // Send remote id
+  int msg_count = 0;
+  for (int i = 0; i < remote_ids.size(); ++i) {
+    if (remote_ids[i].size() != 0) {
+      RPCMessage msg;
+      msg.service_id = service_id;
+      msg.msg_seq = msg_seq;
+      msg.client_id = client_id;
+      int lower = i*group_count;
+      int higher = (i+1)*group_count-1;
+#ifndef _WIN32  // windows does not support rand_r()
+      unsigned int seed = 314;
+      int s_id = (rand_r(&seed) % (higher-lower+1))+lower;
+      msg.server_id = s_id;
+#else
+      LOG(FATAL) << "KVStore does not support Windows yet.";
+#endif
+      msg.data = pickle_data;
+      NDArray tensor = dgl::aten::VecToIdArray<int64_t>(remote_ids[i]);
+      msg.tensors.push_back(tensor);
+      SendRPCMessage(msg, msg.server_id);
+      msg_count++;
+    }
+  }
+  local_data_shape[0] = ID_size;
+  NDArray res_tensor = NDArray::Empty(local_data_shape,
+                                      local_data->dtype,
+                                      DLContext{kDLCPU, 0});
+  char* return_data = static_cast<char*>(res_tensor->data);
+  // Copy local data
+#pragma omp parallel for
+  for (int64_t i = 0; i < local_ids.size(); ++i) {
+    memcpy(return_data + local_ids_orginal[i] * row_size,
+           local_data_char + local_ids[i] * row_size,
+           row_size);
```

Comment:
you don't want to add checks like what i did last time? i found the check was very useful for me to debug the problems. there is no guarantee that users will provide the right Ids. If they do, this will result in segfault. Segfault is very difficult to debug for users.

Reply:
we need to carefully check if the user inputs are correct. if not here, we need to check them somewhere else.

Reply:
My concern is that: this for loop is very time-consuming. If we add two checks here, it actually adds two if branch, I think it could bring the overhead because of the branch prediction fault of CPU processor.

Reply:
I add back these checks. I will test the performance in our benchmark.

Reply:
in this case, memory bandwidth is usually the bottleneck. i don't think adding some check will slow down much.